modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/SnapshotCheckProcess.java

```diff
@@ -0,0 +1,549 @@
+
+                IdleVerifyResultV2 chkRes = SnapshotChecker.reduceHashesResults(results0, errors0);
+
+                finished = clusterOpFut.onDone(new SnapshotPartitionsVerifyTaskResult(locRq.metas, chkRes));
+            }
+            else
+                finished = finishClusterFutureWithErr(clusterOpFut, err, errors0);
+
+            if (finished && log.isInfoEnabled())
+                log.info("Snapshot validation process finished, req: " + locRq + '.');
+        });
+    }
+
+    /** */
+    private Throwable stopAndCleanLocRequest(SnapshotCheckProcessRequest locRq, @Nullable Throwable err) {
+        if (err == null && locRq.error() != null)
+            err = locRq.error();
+
+        GridFutureAdapter<?> locWorkingFut = locRq.fut();
+
+        boolean finished = false;
+
+        // Try to stop local working future ASAP.
+        if (locWorkingFut != null)
+            finished = err == null ? locWorkingFut.onDone() : locWorkingFut.onDone(err);
+
+        requests.remove(locRq.snapshotName());
+
+        if (finished && log.isInfoEnabled())
+            log.info("Finished snapshot local validation, req: " + locRq + '.');
+
+        return err;
+    }
+
+    /** */
+    private Map<ClusterNode, Exception> collectErrors(Map<UUID, Throwable> errors, Set<UUID> requiredNodes) {
+        if (errors.isEmpty())
+            return Collections.emptyMap();
+
+        return errors.entrySet().stream().filter(e -> requiredNodes.contains(e.getKey()) && e.getValue() != null)
+            .collect(Collectors.toMap(e -> kctx.cluster().get().node(e.getKey()), e -> asException(e.getValue())));
+    }
+
+    /** */
+    private Map<ClusterNode, Map<PartitionKeyV2, PartitionHashRecordV2>> collectPartsHashes(
+        @Nullable Map<UUID, CheckResultDTO> results,
+        Collection<UUID> requiredNodes
+    ) {
+        if (F.isEmpty(results))
+            return Collections.emptyMap();
+
+        return results.entrySet().stream()
+            .filter(e -> requiredNodes.contains(e.getKey()) && e.getValue() != null)
+            .collect(Collectors.toMap(e -> kctx.cluster().get().node(e.getKey()), e -> e.getValue().partsHashes));
+    }
+
+    /**
+     * @param snpName Snapshot name of the validation process. If {@code null}, ignored.
+     * @param procId  If {@code snpName} is {@code null}, is used to find the operation request.
+     * @return Current snapshot checking request by {@code snpName} or {@code procId}.
+     */
+    private @Nullable SnapshotCheckProcessRequest currentRequest(@Nullable String snpName, UUID procId) {
+        return snpName == null
+            ? requests.values().stream().filter(rq -> rq.requestId().equals(procId)).findFirst().orElse(null)
+            : requests.get(snpName);
+    }
+
+    /** Phase 1 beginning: prepare, collect and check local metas. */
+    private IgniteInternalFuture<CheckResultDTO> prepareAndCheckMetas(SnapshotCheckProcessRequest extReq) {
```

Comment:
Let's add javadocs for `extReq`. I suppose it means `external`. I'm not sure, why it isn't just a `req`?